### PR TITLE
【Hackathon No.19】Add ASGD RFC

### PR DESCRIPTION
**Pull request by tiancaishaonvjituizi**
添加 Hackathon 任务 19 ASGD 的 RFC

### CODE REVIEW

File: rfcs/APIs/20220327_api_design_for_ASGD.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+class paddle.fluid.optimizer.ASGDOptimizer(learning_rate, parameter_list=None, regularization=None, name=None)
+```
+
+和飞桨中其它优化器的风格保持一致。weight_decay 通过 `regularization` 参数设置，支持 L1/L2 正则。而学习率用 LR Scheduler 来控制，不内置在优化器内。并新增一个 LRScheduler 实现 [2] 中提出的学习率更新策略（具体名字可以后续决定），用户也可以通过使用其它 LRScheduler 或者 LambdaLR，自由选择其它的学习率更新策略。
+
+
+
+## 底层OP设计
+
+基本可以仿照飞桨 SGD 优化器的实现，实现 paddle/fluid/operators/optimizers/asgd_op.cc 和相应的 asgd_kernel.h/.cc/.cu。
```
Comment (zhiboniu) [inline rfcs/APIs/20220327_api_design_for_ASGD.md:120]: 对PyTorch的调研很详细，关于OP实现这部分也写一下详细点的思路吧，让其他开发者根据设计文档也有实现的思路。

Comment (tiancaishaonvjituizi) [reply rfcs/APIs/20220327_api_design_for_ASGD.md]: 好的，补充啦

Comment (zhiboniu) [reply rfcs/APIs/20220327_api_design_for_ASGD.md]: 还是太简介啦😂 ，这样对其他开发者起不到实际指导的作用。
希望能够把论文中的思想与代码结合的方式展现，所以最好包含以下两部分吧：
1）说明ASGD的思想方法
2）方法如何在paddle的optimizer中实现

Comment (tiancaishaonvjituizi) [reply rfcs/APIs/20220327_api_design_for_ASGD.md]: 添加了较为详细的伪代码，请再看一下

File: rfcs/APIs/20220327_api_design_for_ASGD.md [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+经过上面的分析可以发现 PyTorch 的实现是很有问题的。在飞桨里它可以以更加优雅和一致的方式实现。
+
+# 五、设计思路与实现方案
+
+## 命名与参数设计
+```python
+class paddle.fluid.optimizer.ASGDOptimizer(learning_rate, parameter_list=None, regularization=None, name=None)
+```
+
+和飞桨中其它优化器的风格保持一致。weight_decay 通过 `regularization` 参数设置，支持 L1/L2 正则。而学习率用 LR Scheduler 来控制，不内置在优化器内。并新增一个 LRScheduler 实现 [2] 中提出的学习率更新策略（具体名字可以后续决定），用户也可以通过使用其它 LRScheduler 或者 LambdaLR，自由选择其它的学习率更新策略。
```
[the remark this answers is not in the document]
Comment (zhiboniu) [reply rfcs/APIs/20220327_api_design_for_ASGD.md:114]: image
原论文[Acceleration of stochastic approximation by averaging](https://xueshu.baidu.com/usercenter/paper/show?paperid=396ab3d48d059f68d267372498976cd3)中的t^-1是否说明参数更新量也随着时间需要逐渐减少

image
bottou-sgd README中也说明了ASGD需要配合学习率的逐渐衰减。
所以我理解这个学习率更新方法应该是ASGD所要求的。（注意这个学习率衰减的原因与深度学习的学习率下降不是一回事，虽然有些LR Schduler也能起到类似的作用，但是彼此是独立的）

[the remark this answers is not in the document]
Comment (zhiboniu) [reply rfcs/APIs/20220327_api_design_for_ASGD.md:114]: 其实我更关心的不是学习率是否衰减的问题，而是我们是否对ASGD方法和论文已经分析透彻了

[the remark this answers is not in the document]
Comment (tiancaishaonvjituizi) [reply rfcs/APIs/20220327_api_design_for_ASGD.md:114]: > 原论文[Acceleration of stochastic approximation by averaging](https://xueshu.baidu.com/usercenter/paper/show?paperid=396ab3d48d059f68d267372498976cd3)中的t^-1是否说明参数更新量也随着时间需要逐渐减少

这段话不是讲 ASGD 的，是说 “SGD 在满足这个条件的情况下也有和 ASGD 一样最佳的收敛速度，但这个条件在实际中很难满足”。

> 注意这个学习率衰减的原因与深度学习的学习率下降不是一回事，虽然有些LR Schduler也能起到类似的作用，但是彼此是独立的

这句话想表达的意思是不是这样：“ASGD 对学习率衰减策略是有理论上的要求的，不满足要求 ASGD 在理论上就不会有效果” 。

这句话是对的，ASGD 原论文对学习率有一些假设：
    
image

image

但可以看到这些假设是比较宽泛的，可以构造出无数种具体的衰减策略满足这些假设。此外，不满足这些假设的学习率衰减策略就需要被禁止吗？我认为不是的，研究者可能就是想研究 Averaged SGD 在某些特殊条件下的表现。这篇文章 https://arxiv.org/abs/1107.2490 就提到了两个使用固定学习率的研究：

image

因此，不管是从 ASGD 算法本身，还是从 API 的正交性、可组合性上来说，我认为按照 LR Scheduler + 一个不对学习率有任何假设的 Averaged SGD 的思路来实现都是更好的做法。

再附上 sklearn 的 SGD 实现作为另一个旁证：https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html#sklearn.linear_model.SGDClassifier ，可以看到 `average` 是和 `learning_rate`、`penalty` 是正交的参数。

[the remark this answers is not in the document]
Comment (zhiboniu) [reply rfcs/APIs/20220327_api_design_for_ASGD.md:114]: OK

**Review by zhiboniu — APPROVED** [review]
